include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,506 @@
+                        ::std::size_t __wg_agg_idx = __group_id - 1;
+                        __val_type __agg_collector{};
+
+                        // 2a. Calculate the work group's carry-in value.
+                        if (__local_id == 0 && __wg_agg_idx >= 0)
+                        {
+                            for (int32_t __i = __wg_agg_idx; __i >= 0; --__i)
+                            {
+                                const auto& __wg_aggregate = __partials_acc[__i];
+                                const auto& __b_seg_end = __seg_ends_acc[__i];
+
+                                __agg_collector = __binary_op(__wg_aggregate, __agg_collector);
+
+                                // current aggregate is the last aggregate
+                                if (__b_seg_end)
+                                    break;
+                            }
+                        }
+
+                        __agg_collector = sycl::group_broadcast(__group, __agg_collector);
+
+                        // 2c. Second pass over the keys, reidentifying end segments and applying work group
+                        // aggregates if appropriate.
+                        int __start = __global_id * __vals_per_item;
+                        int __end = sycl::minimum{}(__start + __vals_per_item, __n);
+
+                        int __local_min_key_idx = __n - 1;
+
+                        // Find the smallest index in the work group
+                        for (int32_t __i = __start; __i < __end; ++__i)
+                        {
+                            if (__i == __n - 1 || __keys[__i] != __keys[__i + 1])
+                            {
+                                if (__i < __local_min_key_idx)
+                                    __local_min_key_idx = __i;
+                            }
+                        }
+
+                        int __wg_min_seg_end = sycl::reduce_over_group(__group, __local_min_key_idx, sycl::minimum<>());
+
+                        if (__group_id == 0)
+                            return;
+                        
+                        // apply work group aggregates
+                        for (int32_t __i = __start; __i < sycl::minimum{}(__wg_min_seg_end + 1, __end); ++__i)
+                            __out_values[__i] = __binary_op(__agg_collector, __out_values[__i]);
+                    });
+            })
+            .wait();
+    }
+
+    template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3,
+              typename _BinaryPredicate, typename _BinaryOperator, typename T>
+    void operator()(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values, _Range3&& __out_values,
+                    _BinaryPredicate __binary_pred, _BinaryOperator __binary_op, T init = T{})
+    {
+        if constexpr (__scan_type == scan_type::exclusive)
+        {
+            sycl_exclusive_scan_by_segment(::std::forward<_ExecutionPolicy>(__exec), ::std::forward<_Range1>(__keys),
```

Comment:
The semantics of  the `sycl_exclusive_scan_by_segment `and `sycl_inclusive_scan_by_segment` are up to 99% the same.
It leads that implementations are the same too. And it is what we have in the changes... Can we keep just one scan_by_segment implementation with small customization in a couple of places for inclusive/exclusive case?